ecoscope/io/async_earthranger.py

```diff
@@ -0,0 +1,719 @@
+import json
+import geopandas as gpd
+import pandas as pd
+import asyncio
+from dateutil import parser
+from erclient.client import AsyncERClient
+from ecoscope.io.utils import to_hex
+
+import ecoscope
+
+
+class AsyncEarthRangerIO(AsyncERClient):
+    def __init__(self, sub_page_size=4000, tcp_limit=5, **kwargs):
+        if "server" in kwargs:
+            server = kwargs.pop("server")
+            kwargs["service_root"] = f"{server}/api/v1.0"
+            kwargs["token_url"] = f"{server}/oauth2/token"
+
+        self.sub_page_size = sub_page_size
+        self.tcp_limit = tcp_limit
+        self.event_type_display_values = None
+
+        kwargs["client_id"] = kwargs.get("client_id", "das_web_client")
+        super().__init__(**kwargs)
+
+    @staticmethod
+    def _clean_kwargs(addl_kwargs={}, **kwargs):
+        for k in addl_kwargs.keys():
+            print(f"Warning: {k} is a non-standard parameter. Results may be unexpected.")
+        return {k: v for k, v in {**addl_kwargs, **kwargs}.items() if v is not None}
+
+    @staticmethod
+    def _normalize_column(df, col):
+        print(col)
+        for k, v in pd.json_normalize(df.pop(col), sep="__").add_prefix(f"{col}__").items():
+            df[k] = v.values
+
+    @staticmethod
+    def _dataframe_to_dict(events):
+        if isinstance(events, gpd.GeoDataFrame):
+            events["location"] = pd.DataFrame({"longitude": events.geometry.x, "latitude": events.geometry.y}).to_dict(
+                "records"
+            )
+            del events["geometry"]
+
+        if isinstance(events, pd.DataFrame):
+            events = events.to_dict("records")
+        return events
+
+    @staticmethod
+    def _to_gdf(df):
+        longitude, latitude = (0, 1) if isinstance(df["location"].iat[0], list) else ("longitude", "latitude")
+        return gpd.GeoDataFrame(
+            df,
+            geometry=gpd.points_from_xy(df["location"].str[longitude], df["location"].str[latitude]),
+            crs=4326,
+        )
+
+    async def get_sources(
+        self,
+        manufacturer_id=None,
+        provider_key=None,
+        provider=None,
+        id=None,
+        **addl_kwargs,
+    ):
+        """
+        Parameters
+        ----------
+        manufacturer_id
+        provider_key
+        providers
+        id
+        Returns
+        -------
+        An async generator to iterate over sources.
+        """
+
+        params = self._clean_kwargs(
+            addl_kwargs,
+            manufacturer_id=manufacturer_id,
+            provider_key=provider_key,
+            provider=provider,
+            id=id,
+            page_size=addl_kwargs.pop("page_size", self.sub_page_size),
+        )
+
+        async for source in self._get_data("sources/", params=params):
+            yield source
```

Comment:
why do we need this for loop? Can we just yield the response of `_get_data`?

Reply:
Yielding `_get_data` directly yields a generator of a generator, this way we're yielding the json responses as they come in which I think is more intuitive.